### ISSUE TITLE
Adding a package adds it to pixi.toml even if it failed

### ISSUE THREAD

**Issue by AlexisTM** [prefix-dev/pixi#2516]
### Checks

- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the [latest version](https://github.com/prefix-dev/pixi/releases) of pixi, using `pixi --version`.

### Reproducible example

```shell
pixi init
pixi add colcon # results as: Cannot solve the request because of: No candidates were found for colcon *.
pixi add cmake # results as: Cannot solve the request because of: No candidates were found for colcon *. 
```

### Issue description

When adding a package, we may have had a typo or maybe we added a non-existent package.
Regardless, this package will be added to the toml and will make it fail for all further calls.

### Expected behavior

If the package doesn't exist, it shouldn't be added to the toml and report an error the package doesn't exist.

**Comment by pya** [comment]
This looks like  a bug to me. Adding a non-existing package to `[dependencies]` in `pixi.toml` doesn't make any sense. Or, is there a use case where this behavior could be useful? While the fix is simple (Just delete the added line with the package name from `pixi.toml`.), the error message can be very confusing if the typo is subtle.  I just spent several minutes after a typo in a package name to figure out the real cause for repeated error messages after correcting the package name.

Tested with pixi version 0.38.0.

**Comment by ruben-arts** [comment]
You are correct, If I remember correctly we decided to do this as users got frustrated by the solve not working while "importing" or adding a big set of dependencies. They wanted pixi to do the automation of adding it through cli and then edit by hand. 

But this might have been a wrong decision. Could you answer these questions for me?:
- Should it add anything if it doesn't solve correctly? e.g. `pixi add numpy matplotlib python>3.25` could add numpy and matplotlib.
- Should there be a difference if the package name is existing vs non existing in the repodata?
- Should you be able to overwrite this behavior though configuration and or CLI?

**Comment by pya** [comment]
In my opinion, it is important to make it explizit that something was added to the `pixi.toml` that needs manual editing. This could be the "rules":

1. Per default, don't add anything that does not resolve.
2. If configured via CLI or configuration, add the package but also point to the fact that it was not found but added:
    a.  in the error message and
    b.  in  the `pixi.toml` with a comment: 
       `not_found_package = "*"  # package name was not found, package version was not found, package was not found for specified Python version, etc.`
5. If multiple packages are installed and not all resolve, install only the found ones. Maybe an option such as `--atomic-install` could require to install all or nothing?

This topic can get rather complex. But an explaining message can go a long way. Currently, the install is "broken" until the `pixi.toml` is edited.

**Comment by ywilke** [comment]
I asked ChatGPT to generate an overview how different package managers handle the situation where not all the requested packages can be installed. 

| **Package Manager** | **Default Behavior**                 | **Workaround**                                    |
|----------------------|--------------------------------------|--------------------------------------------------|
| `apt`               | Abort all                           | Install individually or use `--fix-broken`      |
| `yum`/`dnf`         | Abort all                           | Use `--skip-broken`                             |
| `zypper`            | Abort all                           | Use `--no-recommends` or install individually   |
| `pacman`            | Abort all                           | Install individually                            |
| `brew`              | Proceed with successful installs    | No workaround needed                           |
| `pip`               | Proceed with successful installs    | No workaround needed                           |
| `conda`             | Abort all                           | Install individually                            |
| `cargo`             | Abort all                           | Install individually                            |
| `npm`               | Proceed with successful installs    | No workaround needed                           |
| `gem`               | Proceed with successful installs    | No workaround needed                           |


> Should it add anything if it doesn't solve correctly? e.g. `pixi add numpy matplotlib python>3.25` could add numpy and matplotlib.

As you can see some package managers abort the installation if one package can't be installed. Not sure if there is a right or wrong choice here. To me it makes more sense to fail if one of the requested packages could not be installed.
> Should there be a difference if the package name is existing vs non existing in the repodata?

It would be nice if the error indicates why a package could not be installed. Either because it didn't exist or because the requested version could not be installed. According to ChatGPT most package managers have a different error message for each situation.

> Should you be able to overwrite this behavior though configuration and or CLI?

The `yum --skip-broken` seems like a nice option to indicate it can skip packages which could not be installed. Could also be called `--skip-failed`.

**Comment by HDembinski** [comment]
Hey, I just came to pixi and I think it is great! The right mix of venv, pip, and conda features, love it, and the seamless PyPI integration is a killer feature.

But the behavior that OP described is really annoying. I want to install as many conda packages as possible with pixi, but I regularly have to fall back to installing from pypi, so it happens to me constantly that I have to edit the pixi.toml to clean up.

**Comment by corneliusroemer** [comment]
I also really like pixi except that it's super annoying I need to remove things manually when I made a typo or guessed wrongly, example session showing workaround:

```
$ pixi add sentence_transformers
  × failed to solve the conda requirements of 'default' 'osx-arm64'
  ╰─▶ Cannot solve the request because of: No candidates were found for sentence_transformers *.

$ pixi add sentence-transformers
  × failed to solve the conda requirements of 'default' 'osx-arm64'
  ╰─▶ Cannot solve the request because of: No candidates were found for sentence_transformers *.

pixi remove sentence_transformers
✔ Removed sentence_transformers
```

**Comment by liquidcarbon** [comment]
+1 to all this - not sure what the right logic for half-failed commands should be, but the current situation is not optimal where once `pixi add` fails, no pixi command works until you `pixi remove` the offender or edit `pixi.toml`

one idea:
for `pixi add thing1 thing2 thing-that-breaks`, break them up internally into separate, single `pixi add` calls, so that the offender can be clearly isolated and never end up on `pixi.toml`

**Comment by AlexisTM** [comment]
I am personally for a `--force` or similar flag for the current behaviour, but fail by default if it would make an invalid workspace.

**Comment by corneliusroemer** [comment]
I agree that it would make sense to add something like `--force`. Better to have a breaking change relatively early on (hopefully) in pixi's lifespan than bake in bad DX.

> You are correct, If I remember correctly we decided to do this as users got frustrated by the solve not working while "importing" or adding a big set of dependencies. They wanted pixi to do the automation of adding it through cli and then edit by hand.

@ruben-arts could you possibly add a link to the discussion/PR that introduced the current behavior to be able to see what the reasoning was back then - out of curiosity mostly, but I think it's good to have things like this linked together. I tried my best to locate an issue/PR but failed to find anything within a few minutes.

**Comment by seembha** [comment]
Thank you for Pixi! I'm just loving it - this is exactly what I was looking for to manage my python projects. If you are still polling for what the behavior should be, this is my personal preference

> - Should it add anything if it doesn't solve correctly? e.g. pixi add numpy matplotlib python>3.25 could add numpy and matplotlib.

This should behave the same way as the case when there are no issues with package names, specs and availability, but a solution doesn't exist. An appropriate message could be printed to stderr or an optional log file.

> - Should there be a difference if the package name is existing vs non existing in the repodata?

No, the behavior should be the same. I prefer fewer exceptions.

> - Should you be able to overwrite this behavior though configuration and or CLI?

Maybe, if the same option is available for the case when there are no name/spec/availability issues, but a solution doesn't exist. In that case how should an "optimal" solution be defined - the one with most compatible packages or should it be based on a priority assigned to packages by their list position (arbitrary, but the channel list in the manifest file sets the precedent). Probably, the latter because the complexity of former can easily get out of hand if the list is large.
